panther_analysis_tool/testing.py

```diff
@@ -41,12 +41,15 @@ class FunctionTestResult:
     # error contains a TestError instance with the error message or
     # None if no error was raised
     error: Optional[TestError]
+    # return value matched expectations
+    matched: Optional[bool]
```

Comment:
looking for input on this approach for checking per-function matched value (helps in deciding to print the policy/rule output since the overall `passed` value can be "false" even if the rule itself has passed, for aux function errors, etc.)

Reply:
Isn't this addressed in https://github.com/panther-labs/panther_analysis_tool/pull/132 ?

Reply:
Well, this specifically allows us to also compare the output of the function with the expected ([here](https://github.com/panther-labs/panther_analysis_tool/pull/133/files#diff-4c3406e46ca13890fa64d295f86ed01e0b0bcd3d28031f1fc97978bb8fca2da3R1199)).  The PR you link allows us to determine whether or not the overall status of the test pass/failed.  But this allows us to print the output when it doesn't match the expected.  Today, this is only useful for the detections themselves, but I could see it being useful for the other functions in the future too (e.g. allow people to specify an expected output from the title function? so we could say the title function failed if there is an error _or_ the expected output doesn't match) Specifically allows us to print this: 

```
AWS.CloudTrail.ResourceMadePublic
	[FAIL] ECR Made Public
		[FAIL] [rule] ZeroDivisionError: division by zero
		[PASS] [dedup] defaultDedupString:AWS.CloudTrail.ResourceMadePublic
		[FAIL] [alertContext] ZeroDivisionError: division by zero
		[FAIL] [description] ZeroDivisionError: division by zero
```

rather than this:

```
AWS.CloudTrail.ResourceMadePublic
	[FAIL] ECR Made Public
		[FAIL] [alert_context] division by zero
		[FAIL] [description] division by zero
```

Reply:
I guess what was addressed in the other PR is being able to tell if the overall test pass/failed.  The test result has these fields:

```python
    detectionId: Optional[str]
    genericError: Optional[str]
    error: Optional[TestError]
    errored: bool
    passed: bool
    trigger_alert: Optional[bool]
    functions: TestResultsPerFunction
```

Which doesn't necessarily tell us if the detection output == expectations.  Since we don't know that from these fields, I'm unsure how we would determine if we should print the error/output from a detection?

Reply:
> but I could see it being useful for the other functions in the future too (e.g. allow people to specify an expected output from the title function? 

Ah nice, it would be useful in this case!

Reply:
This is going to be extraordinarily helpful in debugging rules, very nice

Reply:
The ability to test the output of generated alert fields is an fx request from some customers

https://app.asana.com/0/1200908948600035/1200945739799345/f